File: src/runtime/systemCredentialStore.js
import { isDeepStrictEqual } from 'node:util';
import { randomUUID } from 'node:crypto';
import { constants as fsConstants, promises as defaultFs } from 'node:fs';
import path from 'node:path';
import { validatePortableRecord } from './portableStore.js';
import {
  commandFailure,
  createCommandRunner,
  runCommand,
  SystemCredentialStoreError,
  verifyRoundTrip,
} from './systemCredentialSupport.js';

export { createCommandRunner, SystemCredentialStoreError } from './systemCredentialSupport.js';

const SERVICE_NAME = 'TokensMind Agent Network';
const APPLICATION_NAME = 'tokensmind-agent-network';
const RECORD_NAMES = new Set(['active', 'pending', 'instance']);
const MACOS_SECURITY_PATH = '/usr/bin/security';
const MACOS_ITEM_NOT_FOUND = 44;
const SYSTEM_STORE_PROBE_LENGTH = 1024;

function validateRecordName(name) {
  if (!RECORD_NAMES.has(name)) throw new Error(`Unsupported portable state record: ${name}`);
}

function credentialAccount(namespace, name) {
  validateRecordName(name);
  return `${namespace}:${name}`;
}

function parseRecord(name, source) {
  let value;
  try {
    value = JSON.parse(source);
  } catch (error) {
    throw new SystemCredentialStoreError(
      `System Agent Network ${name} credential is invalid JSON`,
      { cause: error },
    );
  }
  try {
    validatePortableRecord(name, value);
  } catch (error) {
    throw new SystemCredentialStoreError(
      `System Agent Network ${name} credential has an invalid structure`,
      { cause: error },
    );
  }
  return value;
}

function macPasswordInput(source) {
  return `${source}\n${source}\n`;
}

function createProbeValue() {
  return randomUUID().padEnd(SYSTEM_STORE_PROBE_LENGTH, 'x');
}

function createMacCommands(run) {
  const backend = 'macOS Keychain';
  async function readSource(account) {
    const result = await runCommand({
      run, backend, operation: 'read',
      request: {
        command: MACOS_SECURITY_PATH,
        args: ['find-generic-password', '-a', account, '-s', SERVICE_NAME, '-w'],
      },
    });
    if (result.code === MACOS_ITEM_NOT_FOUND) return null;
    if (result.code !== 0) throw commandFailure(backend, 'read', result);
    return result.stdout.trim();
  }
  async function writeSource(account, source) {
    const result = await runCommand({
      run, backend, operation: 'write',
      request: {
        command: MACOS_SECURITY_PATH,
        args: ['add-generic-password', '-a', account, '-s', SERVICE_NAME, '-U', '-w'],
        input: macPasswordInput(source),
      },
    });
    if (result.code !== 0) throw commandFailure(backend, 'write', result);
  }
  async function removeAccount(account) {
    const result = await runCommand({
      run, backend, operation: 'remove',
      request: {
        command: MACOS_SECURITY_PATH,
        args: ['delete-generic-password', '-a', account, '-s', SERVICE_NAME],
      },
    });
    if (![0, MACOS_ITEM_NOT_FOUND].includes(result.code)) {
      throw commandFailure(backend, 'remove', result);
    }
  }
  return { readSource, writeSource, removeAccount };
}

function createMacOperations({ namespace, run }) {
  const { readSource, writeSource, removeAccount } = createMacCommands(run);
  async function readRecord(name) {
    const source = await readSource(credentialAccount(namespace, name));
    return source === null ? null : parseRecord(name, source);
  }
  return {
    read: readRecord,
    async write(name, value) {
      validateRecordName(name);
      validatePortableRecord(name, value);
      const account = credentialAccount(namespace, name);
      await writeSource(account, JSON.stringify(value));
      const stored = await readRecord(name);
      if (!isDeepStrictEqual(stored, value)) {
        throw new SystemCredentialStoreError(`macOS Keychain ${name} write verification failed`);
      }
    },
    async remove(name) {
      await removeAccount(credentialAccount(namespace, name));
    },
    async verifyAvailability() {
      const account = `${namespace}:probe:${randomUUID()}`;
      const expected = createProbeValue();
      await verifyRoundTrip({
        expected,
        write: (source) => writeSource(account, source),
        read: () => readSource(account),
        remove: () => removeAccount(account),
      });
    },
    backend: 'macos-keychain',
  };
}

function linuxAttributes(namespace, name) {
  validateRecordName(name);
  return ['application', APPLICATION_NAME, 'origin', namespace, 'record', name];
}

function createLinuxCommands({ command, run }) {
  const backend = 'Linux Secret Service';
  async function lookupSource(attributes) {
    const result = await runCommand({
      run, backend, operation: 'read',
      request: { command, args: ['lookup', ...attributes] },
    });
    const missing = result.code === 1 && !result.stdout.trim() && !result.stderr.trim();
    if (missing) return null;
    if (result.code !== 0) throw commandFailure(backend, 'read', result);
    return result.stdout.trim();
  }
  async function writeSource(attributes, source) {
    const result = await runCommand({
      run, backend, operation: 'write',
      request: {
        command,
        args: ['store', `--label=${SERVICE_NAME}`, ...attributes],
        input: `${source}\n`,
      },
    });
    if (result.code !== 0) throw commandFailure(backend, 'write', result);
  }
  async function removeAttributes(attributes) {
    if (await lookupSource(attributes) === null) return;
    const result = await runCommand({
      run, backend, operation: 'remove',
      request: { command, args: ['clear', ...attributes] },
    });
    if (result.code !== 0) throw commandFailure(backend, 'remove', result);
  }
  return { lookupSource, writeSource, removeAttributes };
}

function createLinuxOperations({ namespace, command, run }) {
  const { lookupSource, writeSource, removeAttributes } = createLinuxCommands({ command, run });
  async function readRecord(name) {
    const source = await lookupSource(linuxAttributes(namespace, name));
    return source === null ? null : parseRecord(name, source);
  }
  return {
    read: readRecord,
    async write(name, value) {
      validatePortableRecord(name, value);
      const attributes = linuxAttributes(namespace, name);
      await writeSource(attributes, JSON.stringify(value));
      if (!isDeepStrictEqual(await readRecord(name), value)) {
        throw new SystemCredentialStoreError(`Linux Secret Service ${name} write verification failed`);
      }
    },
    async remove(name) {
      await removeAttributes(linuxAttributes(namespace, name));
    },
    async verifyAvailability() {
      const probeName = `probe-${randomUUID()}`;
      const attributes = [
        'application', APPLICATION_NAME, 'origin', namespace, 'record', probeName,
      ];
      const expected = createProbeValue();
      await verifyRoundTrip({
        expected,
        write: (source) => writeSource(attributes, source),
        read: () => lookupSource(attributes),
        remove: () => removeAttributes(attributes),
      });
    },
    backend: 'linux-secret-service',
  };
}

async function isExecutable(fs, filePath) {
  try {
    await fs.access(filePath, fsConstants.X_OK);
    return true;
  } catch {
    return false;
  }
}

async function findExecutable({ name, env, fs }) {
  const directories = String(env.PATH || '').split(path.delimiter).filter(Boolean);
  for (const directory of directories) {
    const candidate = path.join(directory, name);
    if (await isExecutable(fs, candidate)) return candidate;
  }
  return null;
}

export async function resolveSystemCredentialStore({
  platform,
  namespace,
  env = process.env,
  fs = defaultFs,
  run = createCommandRunner(),
} = {}) {
  if (platform === 'darwin') {
    if (!await isExecutable(fs, MACOS_SECURITY_PATH)) return null;
    return createMacOperations({ namespace, run });
  }
  if (platform === 'linux') {
    const command = await findExecutable({ name: 'secret-tool', env, fs });
    return command ? createLinuxOperations({ namespace, command, run }) : null;
  }
  return null;
}

File: skills/tokensmind-agent-network/scripts/python_runtime/system_credential_store.py
import json
import os
import shutil
import subprocess
import uuid

from .portable_store import RECORD_NAMES, _validate_record

SERVICE_NAME = "TokensMind Agent Network"
APPLICATION_NAME = "tokensmind-agent-network"
MACOS_SECURITY_PATH = "/usr/bin/security"
MACOS_ITEM_NOT_FOUND = 44
SYSTEM_STORE_PROBE_LENGTH = 1024


class SystemCredentialStoreError(RuntimeError):
    code = "SYSTEM_CREDENTIAL_STORE_UNAVAILABLE"


def _validate_record_name(name):
    if name not in RECORD_NAMES:
        raise ValueError("Unsupported portable state record: %s" % name)


def _parse_record(name, source):
    try:
        value = json.loads(source)
    except json.JSONDecodeError as error:
        raise SystemCredentialStoreError(
            "System Agent Network %s credential is invalid JSON" % name,
        ) from error
    try:
        _validate_record(name, value)
    except ValueError as error:
        raise SystemCredentialStoreError(
            "System Agent Network %s credential has an invalid structure" % name,
        ) from error
    return value


def _command_failure(backend, operation, result):
    return SystemCredentialStoreError(
        "%s %s failed with exit code %s" % (backend, operation, result.returncode),
    )


def run_credential_command(args, input_text=""):
    return subprocess.run(
        args,
        input=input_text,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
    )


def _run(runner, args, input_text, backend, operation):
    try:
        return runner(args, input_text)
    except OSError as error:
        raise SystemCredentialStoreError(
            "%s %s failed" % (backend, operation),
        ) from error


def _verify_round_trip(write, read, remove, expected):
    failure = None
    try:
        write(expected)
        if read() != expected:
            failure = SystemCredentialStoreError(
                "System credential store round-trip failed",
            )
    except SystemCredentialStoreError as error:
        failure = error
    try:
        remove()
    except SystemCredentialStoreError as error:
        if failure is None:
            failure = error
    if failure is not None:
        raise failure


def _probe_value():
    return str(uuid.uuid4()).ljust(SYSTEM_STORE_PROBE_LENGTH, "x")


class MacOSCredentialStore:
    def __init__(self, namespace, runner=run_credential_command):
        self._namespace = namespace
        self._runner = runner

    def _account(self, name):
        _validate_record_name(name)
        return "%s:%s" % (self._namespace, name)

    def _read_source(self, account):
        result = _run(self._runner, [
            MACOS_SECURITY_PATH, "find-generic-password",
            "-a", account, "-s", SERVICE_NAME, "-w",
        ], "", "macOS Keychain", "read")
        if result.returncode == MACOS_ITEM_NOT_FOUND:
            return None
        if result.returncode != 0:
            raise _command_failure("macOS Keychain", "read", result)
        return result.stdout.strip()

    def _write_source(self, account, source):
        result = _run(self._runner, [
            MACOS_SECURITY_PATH, "add-generic-password",
            "-a", account, "-s", SERVICE_NAME, "-U", "-w",
        ], "%s\n%s\n" % (source, source), "macOS Keychain", "write")
        if result.returncode != 0:
            raise _command_failure("macOS Keychain", "write", result)

    def _remove_account(self, account):
        result = _run(self._runner, [
            MACOS_SECURITY_PATH, "delete-generic-password",
            "-a", account, "-s", SERVICE_NAME,
        ], "", "macOS Keychain", "remove")
        if result.returncode not in (0, MACOS_ITEM_NOT_FOUND):
            raise _command_failure("macOS Keychain", "remove", result)

    def read(self, name):
        source = self._read_source(self._account(name))
        return None if source is None else _parse_record(name, source)

    def write(self, name, value):
        _validate_record(name, value)
        self._write_source(
            self._account(name), json.dumps(value, separators=(",", ":")),
        )
        if self.read(name) != value:
            raise SystemCredentialStoreError(
                "macOS Keychain %s write verification failed" % name,
            )

    def remove(self, name):
        self._remove_account(self._account(name))

    def verify_availability(self):
        account = "%s:probe:%s" % (self._namespace, uuid.uuid4())
        expected = _probe_value()
        _verify_round_trip(
            lambda source: self._write_source(account, source),
            lambda: self._read_source(account),
            lambda: self._remove_account(account),
            expected,
        )

    backend = "macos-keychain"


class LinuxSecretServiceStore:
    def __init__(self, namespace, command, runner=run_credential_command):
        self._namespace = namespace
        self._command = command
        self._runner = runner

    def _attributes(self, name):
        _validate_record_name(name)
        return [
            "application", APPLICATION_NAME,
            "origin", self._namespace,
            "record", name,
        ]

    def _lookup_source(self, attributes):
        result = _run(
            self._runner, [self._command, "lookup"] + attributes,
            "", "Linux Secret Service", "read",
        )
        missing = (
            result.returncode == 1
            and not result.stdout.strip()
            and not result.stderr.strip()
        )
        if missing:
            return None
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "read", result)
        return result.stdout.strip()

    def _write_source(self, attributes, source):
        result = _run(self._runner, [
            self._command, "store", "--label=%s" % SERVICE_NAME,
        ] + attributes, source + "\n", "Linux Secret Service", "write")
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "write", result)

    def _remove_attributes(self, attributes):
        if self._lookup_source(attributes) is None:
            return
        result = _run(
            self._runner, [self._command, "clear"] + attributes,
            "", "Linux Secret Service", "remove",
        )
        if result.returncode != 0:
            raise _command_failure("Linux Secret Service", "remove", result)

    def read(self, name):
        source = self._lookup_source(self._attributes(name))
        return None if source is None else _parse_record(name, source)

    def write(self, name, value):
        _validate_record(name, value)
        self._write_source(
            self._attributes(name), json.dumps(value, separators=(",", ":")),
        )
        if self.read(name) != value:
            raise SystemCredentialStoreError(
                "Linux Secret Service %s write verification failed" % name,
            )

    def remove(self, name):
        self._remove_attributes(self._attributes(name))

    def verify_availability(self):
        attributes = [
            "application", APPLICATION_NAME,
            "origin", self._namespace,
            "record", "probe-%s" % uuid.uuid4(),
        ]
        expected = _probe_value()
        _verify_round_trip(
            lambda source: self._write_source(attributes, source),
            lambda: self._lookup_source(attributes),
            lambda: self._remove_attributes(attributes),
            expected,
        )

    backend = "linux-secret-service"


def resolve_system_credential_store(
        platform, namespace, env=None, *, runner=run_credential_command):
    if platform == "darwin":
        if not os.path.isfile(MACOS_SECURITY_PATH):
            return None
        if not os.access(MACOS_SECURITY_PATH, os.X_OK):
            return None
        return MacOSCredentialStore(namespace, runner)
    if platform == "linux":
        environment = os.environ if env is None else env
        command = shutil.which("secret-tool", path=environment.get("PATH"))
        if command:
            return LinuxSecretServiceStore(namespace, command, runner)
    return None
